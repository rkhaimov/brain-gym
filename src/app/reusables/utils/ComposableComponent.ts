import { OnChanges, OnDestroy, OnInit, SimpleChanges } from '@angular/core';
import { identity, noop } from 'rxjs';

function Identity() {
  return identity;
}

export type Hook = OnDestroy & OnInit & OnChanges;

export function createHook(config: Partial<Hook>): Hook {
  return {
    ngOnChanges: noop,
    ngOnInit: noop,
    ngOnDestroy: noop,
    ...config,
  };
}

export function joinHooks(left: Hook, right: Hook): Hook {
  return createHook({
    ngOnChanges: (changes) => {
      left.ngOnChanges(changes);
      right.ngOnChanges(changes);
    },
    ngOnInit: () => {
      left.ngOnInit();
      right.ngOnInit();
    },
    ngOnDestroy: () => {
      left.ngOnDestroy();
      right.ngOnDestroy();
    },
  });
}

@Identity()
export class ComposableComponent implements OnDestroy, OnInit, OnChanges {
  private fns: Hook[] = [];

  ngOnInit() {
    this.fns.forEach((fn) => 'ngOnInit' in fn && fn.ngOnInit());
  }

  ngOnDestroy() {
    this.fns.forEach((fn) => 'ngOnDestroy' in fn && fn.ngOnDestroy());
  }

  ngOnChanges(changes: SimpleChanges) {
    this.fns.forEach((fn) => 'ngOnChanges' in fn && fn.ngOnChanges(changes));
  }

  protected compose(...fns: Hook[]): void {
    this.fns = fns;
  }
}
