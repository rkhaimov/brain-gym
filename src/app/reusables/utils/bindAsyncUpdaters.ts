import { merge, Observable, switchMap, tap, withLatestFrom } from 'rxjs';
import { Storage } from './createStorage';

export type AsyncUpdater<TState> = (state: TState) => Observable<TState>;

export const bindAsyncUpdaters = <TState>(
  updaters: Observable<AsyncUpdater<TState>>[],
  storage: Storage<TState>
): Observable<unknown> =>
  merge(...updaters).pipe(
    withLatestFrom(storage.state$),
    switchMap(([updater, state]) => updater(state)),
    tap((state) => storage.update(() => state))
  );
