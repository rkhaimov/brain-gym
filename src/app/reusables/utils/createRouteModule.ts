import { NgModule } from '@angular/core';
import { Route, RouterModule } from '@angular/router';

export function createRouteModule(route: Route, config: NgModule): Route {
  @NgModule({
    ...config,
    imports: [
      RouterModule.forChild([{ path: '', component: route.component }]),
    ],
  })
  class Inner {}

  return { path: route.path, loadChildren: () => Promise.resolve(Inner) };
}
