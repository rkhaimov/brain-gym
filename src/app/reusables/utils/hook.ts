import { noop } from 'rxjs';
import { OnChanges, OnDestroy, OnInit } from '@angular/core';

export type Hook = OnDestroy & OnInit & OnChanges;

export function createHook(config: Partial<Hook>): Hook {
  return {
    ngOnChanges: noop,
    ngOnInit: noop,
    ngOnDestroy: noop,
    ...config,
  };
}

export function joinHooks(left: Hook, right: Hook): Hook {
  return createHook({
    ngOnChanges: (changes) => {
      left.ngOnChanges(changes);
      right.ngOnChanges(changes);
    },
    ngOnInit: () => {
      left.ngOnInit();
      right.ngOnInit();
    },
    ngOnDestroy: () => {
      left.ngOnDestroy();
      right.ngOnDestroy();
    },
  });
}
