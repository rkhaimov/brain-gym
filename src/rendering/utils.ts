import { prettyDOM } from '@testing-library/dom';
import { last } from 'lodash-es';
import ReactDOM from 'react-dom';
import { Observable } from 'rxjs';
import { VirtualNodeOrString } from './createElement';
import { createRenderer } from './render';

export function renderReact(jsx: JSX.Element) {
  const container = document.createElement('main');

  ReactDOM.render(jsx, container);

  return {
    container,
    async act(clb: () => unknown) {
      document.body.appendChild(container);

      await clb();

      document.body.removeChild(container);
    },
    get output() {
      return prettyDOM(container);
    },
  };
}

export function renderObservable(element: Observable<VirtualNodeOrString>) {
  let outputs: string[] = [];
  let error: Error;
  const container = document.createElement('main');
  const render = createRenderer();

  const subscription = element.subscribe((ui) => {
    try {
      outputs.push(prettyDOM(render(ui, container)) as string);
    } catch (e: unknown) {
      error = e as Error;

      subscription.unsubscribe();
    }
  });

  return {
    get output() {
      if (error) {
        throw error;
      }

      return last(outputs);
    },
    container,
    async act(clb: () => unknown) {
      document.body.appendChild(container);

      await clb();

      document.body.removeChild(container);
    },
  };
}
