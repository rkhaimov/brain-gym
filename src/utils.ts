import { Point } from './Point';

export function assertNotUndefined<T>(
  value: T
): asserts value is NonNullable<T> {
  assert(value !== undefined);
}

export function assert(
  condition: any,
  message = 'Condition is false',
  onError = noop,
): asserts condition {
  if (condition === false) {
    onError();

    throw new AssertionError(message);
  }
}

class AssertionError extends Error {
  constructor(message: string) {
    super(`Assertion has failed: ${message}`);
  }
}

export function sequenceLoop(
  points: Point[],
  index: number,
  animation: (from: Point, to: Point, onEnd: () => void) => void
) {
  const next = (index + 1) % points.length;

  assertNotUndefined(points[next]);

  animation(points[index], points[next], () =>
    sequenceLoop(points, next, animation)
  );
}

function noop() {

}
