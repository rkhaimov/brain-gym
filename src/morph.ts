import { moveInALine } from './moves';
import { Point } from './Point';
import { assert } from './utils';

export function morphInTo(from: SVGShape, to: Shape, onEnd: () => void) {
  const moves = getPointsMoves(from, to);
  const change = interpolate(0, 1, 1_000);

  requestAnimationFrame(update);

  function update() {
    change.perform();

    if (change.complete()) {
      onEnd();

      return;
    }

    for (const move of moves) {
      moveInALine(move.from.point, move.to.point, change.current);
    }

    const path = moves.map((move) => `${move.type}${move.from.point.position.x} ${move.from.point.position.y}`).join(' ');

    from.setAttribute('d', `${path} Z`);

    requestAnimationFrame(update);
  }
}

export function getPointsMoves(from: Shape, to: Shape) {
  const fromPath = parsePath(from.getAttribute('d'));
  const toPath = parsePath(to.getAttribute('d'));
  const diff = toPath.length - fromPath.length;

  if (diff > 0) {
    equalize(diff, fromPath);
  }

  if (diff < 0) {
    equalize(diff * -1, toPath);
  }

  return buildTransformations(fromPath, toPath);
}

function equalize(diff: number, toFill: ReturnType<typeof parsePath>) {
  assert(diff > 0);

  for (let index = 0; index < diff; index += 1) {
    toFill.push({ type: 'L', position: toFill[toFill.length - 1].position });
  }
}

function parsePath(path: string | null): PathPoint[] {
  assertPathIsValid(path);

  const match = path.match(/[A-Z]+\d+ \d+/g);

  assert(match !== null);

  return match.map((point) => {
    const [x, y] = point.slice(1).split(' ');

    return {
      type: point[0],
      position: {
        x: parseInt(x, 10),
        y: parseInt(y, 10),
      },
    };
  });
}

function assertPathIsValid(path: string | null): asserts path is string {
  assert(path !== null, 'Path is null');
  assert(path[0] === 'M', 'Path is not starting by M');
  assert(path[path.length - 1] === 'Z', 'Path is not ending by Z');

  const types = new Set(path.slice(1, path.length - 1).match(/[A-Z]+/g));

  assert(
    types.size === 1 && types.has('L'),
    `Path contains unknown chars ${types}. Should consist only of L's`
  );
}

function buildTransformations(fromPath: PathPoint[], toPath: PathPoint[]) {
  assert(fromPath.length === toPath.length);

  return fromPath.reduce((accumulated, _, index) => {
    assert(toPath[index].type === fromPath[index].type)

    accumulated.push({
      type: toPath[index].type,
      from: {
        point: new Point(fromPath[index].position),
      },
      to: {
        point: new Point(toPath[index].position),
      },
    });

    return accumulated;
  }, [] as TransformationMeta[]);
}

type SVGShape = Shape & {
  setAttribute(qualifiedName: string, attr: string): unknown;
}

export type Shape = {
  getAttribute(qualifiedName: string): string | null;
};

type PathPoint = {
  type: string;
  position: Point['position'];
};

type TransformationMeta = {
  type: string;
  from: {
    point: Point;
  };
  to: {
    point: Point;
  };
};

export function interpolate(from: number, to: number, time: number) {
  const FRAME_TIME = 10;
  const diff = (to - from) / (time / FRAME_TIME);

  return {
    current: 0,
    complete() {
      return parseFloat(this.current.toFixed(5)) === to;
    },
    perform() {
      this.current += diff;
    },
  };
}