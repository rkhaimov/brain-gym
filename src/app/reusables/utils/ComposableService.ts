import { OnDestroy } from '@angular/core';
import { identity } from 'rxjs';
import { Hook } from './hook';

function Identity() {
  return identity;
}

@Identity()
export class ComposableService implements OnDestroy {
  private fns: Hook[] = [];

  ngOnDestroy() {
    this.fns.forEach((fn) => fn.ngOnDestroy());
  }

  protected compose(...fns: Hook[]): void {
    fns.forEach((fn) => fn.ngOnInit());

    this.fns = fns;
  }
}
