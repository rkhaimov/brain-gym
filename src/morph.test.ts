import { getPointsMoves, Shape } from './morph';
import { assert } from './utils';

describe('morph', () => {
  it('should morph shapes of equal sizes', () => {
    const s1 = createShape('M0 0 L5 5 L10 30 Z');
    const s2 = createShape('M10 0 L15 5 L20 30 Z');

    expect(getPointsMoves(s1, s2)).toMatchSnapshot();
  });

  it('should morph shapes of bigger sizes', () => {
    const s1 = createShape('M0 0 L5 5 L10 30 Z');
    const s2 = createShape('M10 0 L15 5 L20 30 L30 40 L60 40 Z');

    expect(getPointsMoves(s1, s2)).toMatchSnapshot();
  });

  it('should morph shapes of lesser sizes', () => {
    const s1 = createShape('M0 0 L5 5 L10 30 Z');
    const s2 = createShape('M10 0 L15 5 Z');

    expect(getPointsMoves(s1, s2)).toMatchSnapshot();
  });

  it('should not accept invalid paths', () => {
    expect(() =>
      getPointsMoves(
        createShape('0 0 L5 5 L10 30 Z'),
        createShape('M0 0 L0 0 Z')
      )
    ).toThrow();

    expect(() =>
      getPointsMoves(
        createShape('M0 0 L5 5 L10 30'),
        createShape('M0 0 L0 0 Z')
      )
    ).toThrow();

    expect(() =>
      getPointsMoves(
        createShape('M0 0 L5 5 X10 30 Z'),
        createShape('M0 0 L0 0 Z')
      )
    ).toThrow();
  });
});

function createShape(path: string): Shape {
  return {
    getAttribute: (qualifiedName) => {
      assert(qualifiedName === 'd', 'Unknown qualifier');

      return path;
    },
  };
}
