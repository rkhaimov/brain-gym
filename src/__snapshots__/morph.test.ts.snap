// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`morph should morph shapes of bigger sizes 1`] = `
Array [
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 0,
          "y": 0,
        },
      },
      "type": "M",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 0,
        },
      },
      "type": "M",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 5,
          "y": 5,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 15,
          "y": 5,
        },
      },
      "type": "L",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 30,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 20,
          "y": 30,
        },
      },
      "type": "L",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 30,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 30,
          "y": 40,
        },
      },
      "type": "L",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 30,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 60,
          "y": 40,
        },
      },
      "type": "L",
    },
  },
]
`;

exports[`morph should morph shapes of equal sizes 1`] = `
Array [
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 0,
          "y": 0,
        },
      },
      "type": "M",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 0,
        },
      },
      "type": "M",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 5,
          "y": 5,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 15,
          "y": 5,
        },
      },
      "type": "L",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 30,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 20,
          "y": 30,
        },
      },
      "type": "L",
    },
  },
]
`;

exports[`morph should morph shapes of lesser sizes 1`] = `
Array [
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 0,
          "y": 0,
        },
      },
      "type": "M",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 0,
        },
      },
      "type": "M",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 5,
          "y": 5,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 15,
          "y": 5,
        },
      },
      "type": "L",
    },
  },
  Object {
    "from": Object {
      "point": Point {
        "position": Object {
          "x": 10,
          "y": 30,
        },
      },
      "type": "L",
    },
    "to": Object {
      "point": Point {
        "position": Object {
          "x": 15,
          "y": 5,
        },
      },
      "type": "L",
    },
  },
]
`;
